### DISCUSSION TITLE
Missing values

### DISCUSSION THREAD

**Discussion by christopher-wittlinger** [skfolio/skfolio#88] [General]
I am currently trying out skfolio and it is pretty great. I have one issue however. We have a investible universe of a few hundred stocks, not all stocks IPO'd at the same time, therefore when constructing the price dataframe and the corresponding returns dataframe, it includes NaN values.

When runnnig an optimization I get the issue that I should use an imputer transformer. I am not sure how I would achieve my desired result with that. Anyone has an idea?

**Comment by Conway-Yu** [comment]
Hello,

you can find the related discussions here:
https://github.com/skfolio/skfolio/issues/67
https://github.com/skfolio/skfolio/discussions/60

**Reply by christopher-wittlinger** [reply]
@Conway-Yu thanks for your answer - Is this already available though? How would I use it?

**Reply by christopher-wittlinger** [reply]
@Conway-Yu Also if I am not mistaken, isn't a transformer being used to remove data initially? I want to use all stocks in my optimization, but stocks that have not yet started I want to ignore until they started.

**Reply by Conway-Yu** [reply]
As far as I know it is not officially implemented yet. Maybe @HugoDelatte can give some more insights on the roadmap.

I used my quick, simple and imperfect solution as described in (https://github.com/skfolio/skfolio/discussions/60). 

However, this was bond specific and so I can imagine you need to modify the code a bit, namely removing the following part:

```python
        # Remove assets that are set to expire within the next x months in the test set
        if self.expiration_dates is not None:
            for i, expiration_date in enumerate(self.expiration_dates):
                if expiration_date is not None and expiration_date <= (n_timestep + relativedelta(months = self.test_size)):
                    to_keep_[i] = False

        self.to_keep_ = np.array(to_keep_)
```

Regarding your second question, yes you are correct in your understanding.
An option to have your desired behavior is in addition to use a [Walkforward](https://skfolio.org/generated/skfolio.model_selection.WalkForward.html#skfolio.model_selection.WalkForward) analysis. Then whenever a new stock is introduced, it will be considered in the optimizer. Until then its ignored.

Edit: added link to walkforward implementation in skfolio

**Reply by HugoDelatte** [reply]
Yes, that is correct. 
The implementation is next on the roadmap, with an EAT of one month.
I'll let you know when available.

**Reply by christopher-wittlinger** [reply]
@Conway-Yu thanks for the example I will take a look. The Walkforward is also the right concept to use when we want to enforce certain rebalancing behavior (e.g. Rebalance every month / quarter), right?

@HugoDelatte thanks for the answer - to understand you correctly within the next month you will add this functionality on your side, so I can wait for it and use the official solution? Is there anything I can do to facilitate/help with your work?

**Reply by HugoDelatte** [reply]
Correct. I'll show you the implementation when it's done to have your feedback.
Concerning your question on the Walkforward, it's indeed the right tool. We recently released time specific rebalancing with the [freq param](https://skfolio.org/generated/skfolio.model_selection.WalkForward.html#skfolio.model_selection.WalkForward).

**Reply by christopher-wittlinger** [reply]
Thanks. Looking forward.
